File: android-core/src/test/java/com/mparticle/commerce/ProductTest.java
package com.mparticle.commerce;

import com.mparticle.MParticle;
import com.mparticle.MockMParticle;

import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.*;


public class ProductTest {

    @Before
    public void before() {
        MParticle.setInstance(new MockMParticle());
    }

    @Test
    public void testDefaultEqualityComparator() {
        Product.setEqualityComparator(null);
        Product product1 = new Product.Builder("name", "sku",2).brand("cool brand!").build();
        Product product2 = new Product.Builder("cool brand!", "sku",2).brand("cool brand!adsflkjh").build();
        Product product2Copy = new Product.Builder(product2).build();
        product2Copy.mTimeAdded = product2.mTimeAdded;
        assertNotEquals(product2, product1);
        assertEquals(product1, product1);
        assertEquals(product2, product2Copy);

        assertNotEquals(product1, null);
        assertNotEquals(null, product1);
    }

    @Test
    public void testEqualityComparator() {
        Product.setEqualityComparator(new Product.EqualityComparator() {
            @Override
            public boolean equals(Product product1, Product product2) {
                return product1.getName().equals(product2.getBrand());
            }
        });
        Product product1 = new Product.Builder("name", "sku",2).brand("cool brand!").build();
        Product product2 = new Product.Builder("cool brand!", "sku",2).brand("cool brand!adsflkjh").build();
        assertEquals(product2, product1);
    }

    @Test
    public void testSerializationDeserialization() {
        Product.setEqualityComparator(new Product.EqualityComparator() {
            @Override
            public boolean equals(Product product1, Product product2) {
                return product1.toString().equals(product2.toString());
            }
        });
        Product product = new Product.Builder("product name", "product sku", 301.45)
                .brand("product brand")
                .category("product category")
                .couponCode("product coupon code")
                .name("product name")
                .position(4)
                .variant("product variant")
                .quantity(12.1)
                .build();

        JSONObject productJson = product.toJson();
        Product product2 = Product.fromJson(productJson);
        assertEquals(product, product2);
        product2.setQuantity(10000);
        assertNotEquals(product, product2);
    }
}

File: android-core/src/test/kotlin/com/mparticle/commerce/ProductTest.kt
package com.mparticle.commerce

import com.mparticle.MParticle
import com.mparticle.MockMParticle
import org.junit.Assert
import org.junit.Before
import org.junit.Test

class ProductTest {
    @Before
    fun before() {
        MParticle.setInstance(MockMParticle())
    }

    @Test
    fun testDefaultEqualityComparator() {
        Product.setEqualityComparator(null)
        val product1 = Product.Builder("name", "sku", 2.0).brand("cool brand!").build()
        val product2 =
            Product.Builder("cool brand!", "sku", 2.0).brand("cool brand!adsflkjh").build()
        val product2Copy = Product.Builder(product2).build()
        product2Copy.mTimeAdded = product2.mTimeAdded
        Assert.assertNotEquals(product2, product1)
        Assert.assertEquals(product1, product1)
        Assert.assertEquals(product2, product2Copy)
        Assert.assertNotEquals(product1, null)
        Assert.assertNotEquals(null, product1)
    }

    @Test
    fun testEqualityComparator() {
        Product.setEqualityComparator { product1, product2 -> product1?.name == product2?.brand }
        val product1 = Product.Builder("name", "sku", 2.0).brand("cool brand!").build()
        val product2 =
            Product.Builder("cool brand!", "sku", 2.0).brand("cool brand!adsflkjh").build()
        Assert.assertEquals(product2, product1)
    }

    @Test
    fun testSerializationDeserialization() {
        Product.setEqualityComparator { product1, product2 -> product1.toString() == product2.toString() }
        val product = Product.Builder("product name", "product sku", 301.45)
            .brand("product brand")
            .category("product category")
            .couponCode("product coupon code")
            .name("product name")
            .position(4)
            .variant("product variant")
            .quantity(12.1)
            .build()
        val productJson = product.toJson()
        val product2 = Product.fromJson(productJson)
        Assert.assertEquals(product, product2)
        product2?.quantity = 10000.0
        Assert.assertNotEquals(product, product2)
    }
}
